### PR TITLE
brianyin/ajs-88-add-tool-class-generic-tool-typing

### PR DESCRIPTION
**Pull request by toubatbrian**
Setup new tool user-facing APIs and refactored the schema to align with updated interfaces.
- Defined new tool interfaces modeled similar to lk-agent v1.0, Vercel, and Mastra’s tool interface.
- Enabled agent definitions using the new tool interfaces (definition only for now—execution logic will follow in upcoming PRs).
- Refactored parts of the existing implementation from the old function context to the new tool context.

Most of the diffs involve copying existing helper functions, renaming parameters, and structural refactoring. Happy to walk through it if a read-through would help!

### CODE REVIEW

File: agents/src/llm/utils.ts [hunk 1]
```diff
@@ -95,3 +98,157 @@ export async function serializeImage(image: ImageContent): Promise<SerializedIma
...
+  if (!tool) {
+    throw new Error(`AI tool ${toolName} not found`);
+  }
+
+  return FunctionCall.create({
+    callId: toolCallId,
+    name: toolName,
+    args: rawArgs,
+  });
+};
```
Comment (toubatbrian) [inline agents/src/llm/utils.ts]: These functions are copied over from the existing `function_context.ts` file. No need to review them

File: agents/src/llm/index.ts [hunk 2]
```diff
@@ -1,15 +1,25 @@
...
 // SPDX-License-Identifier: Apache-2.0
+// export {
+//   oaiBuildFunctionInfo,
+//   oaiParams,
+//   type CallableFunction,
+//   type CallableFunctionResult,
+//   type FunctionCallInfo,
+//   type FunctionContext,
+//   type inferParameters,
+// } from './function_context.js';
```
Comment (Shubhrakanti) [inline agents/src/llm/index.ts]: why not just delete?

File: agents/src/voice/agent.test.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    const agent = createAgent({ instructions, tools });
+
+    const tools1 = agent.tools;
+    const tools2 = agent.tools;
+
+    // Should return different object references
+    expect(tools1).not.toBe(tools2);
+    expect(tools1).not.toBe(tools);
+
+    // But should have the same content
```
Comment (Shubhrakanti) [inline agents/src/voice/agent.test.ts]: Let's be careful of the AI generated comments.

File: agents/src/voice/agent.test.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    expect(agentTools).toHaveProperty('getTool2');
+
+    // Verify tool properties with proper checks
+    expect(agentTools.getTool1?.name).toBe('getTool1');
+    expect(agentTools.getTool1?.description).toBe('First test tool');
+    expect(agentTools.getTool2?.name).toBe('getTool2');
+    expect(agentTools.getTool2?.description).toBe('Second test tool');
+  });
+
+  it('should return a copy of tools, not the original reference', () => {
```
Comment (Shubhrakanti) [inline agents/src/voice/agent.test.ts:62]: These tests feel really unnecessary. I think we should only have tests for edge cases or describe core behavior of a class. @theomonnom @lukasIO would be good to hear to perspective on tests.

File: agents/src/voice/agent.ts [hunk 5]
```diff
@@ -23,33 +30,42 @@ export class StopResponse extends Error {
...
+  tools?: ToolContext;
+  turnDetection?: TurnDetectionMode;
+  stt?: STT;
+  vad?: VAD;
+  llm?: LLM; // TODO: support realtime model
+  tts?: TTS;
+  mcpServers?: any[]; // TODO: support MCP servers
+  allowInterruptions?: boolean;
+  minConsecutiveSpeechDelay?: number;
+}
```
Comment (Shubhrakanti) [inline agents/src/voice/agent.ts:45]: I like this, but let's try to do this in a separate PR (next time) - esp when this one is getting so big. PRs should aim to do one logical unit of work.

Comment (Shubhrakanti) [reply agents/src/voice/agent.ts:45]: In the future the whole restricting go the Agent class parameters should be one PR - it's just code clean up and has nothing to do with tool calling.

File: agents/src/voice/agent.ts [hunk 6]
```diff
@@ -69,13 +84,18 @@ export class Agent {
...
   }
 
   get instructions(): string {
     return this._instructions;
   }
 
+  get tools(): ToolContext {
+    return { ...this._tools };
+  }
+
```
Comment (Shubhrakanti) [inline agents/src/voice/agent.ts:98]: this the only real `tool` related change for this file.

File: examples/src/basic_agent.ts [hunk 7]
```diff
@@ -20,9 +20,10 @@ export default defineAgent({
     proc.userData.vad = await silero.VAD.load();
   },
   entry: async (ctx: JobContext) => {
-    const agent = new voice.Agent(
-      "You are a helpful assistant, you can hear the user's message and respond to it.",
-    );
+    const agent = new voice.Agent({
+      instructions:
+        "You are a helpful assistant, you can hear the user's message and respond to it.",
+    });
```
Comment (Shubhrakanti) [inline examples/src/basic_agent.ts:26]: Did the linter do this?

File: agents/src/llm/provider_format/openai.test.ts [hunk 8]
```diff
@@ -485,7 +485,7 @@ describe('toChatCtx', () => {
         {
           type: 'audio_content',
           frame: [],
-        } as any,
+        } as any, // eslint-disable-line @typescript-eslint/no-explicit-any
```
Comment (Shubhrakanti) [inline agents/src/llm/provider_format/openai.test.ts]: why do we have `any` in the first place?

**Review by Shubhrakanti — COMMENTED** [review]
The title of the PR is "add-tool-class-generic-tool", but it's doing 3 different things 

> Defined new tool interfaces modeled similar to lk-agent v1.0, Vercel, and Mastra’s tool interface.
Enabled agent definitions using the new tool interfaces (definition only for now—execution logic will follow in upcoming PRs).
Refactored parts of the existing implementation from the old function context to the new tool context.


+ refactors the `Agent.ts` to use options as input.
+ renaming function -> tool

I think with this PR it's _okay_, since it won't break the dev branch. But in the future it's best to keep PRs small. If we introduce a bug - the first thing we want to do is revert - so that no one else is blocked. the smaller the PR the easier that becomes. 

It's also easier to review when a PR does just 1-3 things.

File: agents/src/voice/agent.test.ts (hunk 4 above)
Comment (toubatbrian) [reply agents/src/voice/agent.test.ts:62]: The lk-agent python package returns a copy of the tools, which is why I think it’s important to explicitly list this test.

File: examples/src/basic_agent.ts (hunk 7 above)
Comment (toubatbrian) [reply examples/src/basic_agent.ts:26]: Linter cannot recognize this, I manually changed those lines

File: agents/src/llm/provider_format/openai.test.ts (hunk 8 above)
Comment (toubatbrian) [reply agents/src/llm/provider_format/openai.test.ts]: This should not have linting issue without the comment, I just removed it

File: examples/src/basic_agent.ts (hunk 7 above)
Comment (Shubhrakanti) [reply examples/src/basic_agent.ts:26]: Ah I mistook this for a different issue I had earlier.

File: agents/src/voice/agent.test.ts (hunk 4 above)
Comment (Shubhrakanti) [reply agents/src/voice/agent.test.ts:62]: I guess no second thought it's not a bad idea to have more tests. No real downside.

File: agents/src/voice/agent.ts (hunk 6 above)
Comment (toubatbrian) [reply agents/src/voice/agent.ts:98]: Yep, I wouldn't involve anything else for now, just trying to keep this PR size manageable.

**Review by Shubhrakanti — APPROVED** [review]

File: agents/src/voice/agent.test.ts (hunk 4 above)
Comment (toubatbrian) [reply agents/src/voice/agent.test.ts:62]: I guess one reason behind not returning original reference is to prevent user editing the mutable tool list during runtime, which might create undefined behavior

File: agents/src/voice/agent.ts [hunk 9]
```diff
@@ -23,33 +30,42 @@ export class StopResponse extends Error {
...
 }
 
+export interface AgentOptions {
+  instructions: string;
+  chatCtx?: ChatContext;
+  tools?: ToolContext;
+  turnDetection?: TurnDetectionMode;
+  stt?: STT;
+  vad?: VAD;
+  llm?: LLM; // TODO: support realtime model
```
Comment (Shubhrakanti) [inline agents/src/voice/agent.ts:40]: Can we add ticket numbers to this? (Ex. https://linear.app/livekit/issue/AJS-32/add-support-for-realtime-model)
